warp10/src/main/java/io/warp10/script/functions/MOTIONSPLIT.java

```diff
@@ -32,174 +33,376 @@
-        refLocation = location;
-        refTimestamp = timestamp;
-        proximityZoneTraveledDistance = 0.0D;
-      } else if (GeoTimeSerie.NO_LOCATION != refLocation && GeoTimeSerie.NO_LOCATION != location && GeoXPLib.orthodromicDistance(refLocation, location) <= proximityZoneRadius) {
-        if (GeoTimeSerie.NO_LOCATION != previousLocation) {
-          proximityZoneTraveledDistance += GeoXPLib.orthodromicDistance(previousLocation, location);
+
+        // quit the radius, or speed greater than max stopped speed.
+        if (GeoXPLib.orthodromicDistance(refLocation, location) > proximityZoneRadius || currentSpeed > proximityZoneMaxSpeed) {
+          double zoneMeanSpeed = 0.0D;
+          if (previousTimestamp != refTimestamp) {
+            zoneMeanSpeed = proximityZoneTraveledDistance / ((double) (previousTimestamp - refTimestamp) / Constants.TIME_UNITS_PER_S);
+          }
+          if (timeStopped > proximityZoneTime && zoneMeanSpeed < proximityInZoneMaxMeanSpeed) {
+            splitReason = SPLIT_TYPE_STOPPED;
+            mustSplit = true;
+            if (null != stoppedTimeLabel && null != split) {
+              split.getMetadata().putToLabels(stoppedTimeLabel, Long.toString(timeStopped));
+            }
+          }
+          refLocation = location;
+          refTimestamp = timestamp;
+          proximityZoneTraveledDistance = 0.0D;
+        } else { // inside the radius
+          if (GeoTimeSerie.NO_LOCATION != previousLocation) {
+            proximityZoneTraveledDistance += GeoXPLib.orthodromicDistance(previousLocation, location);
+          }
+        }
+      }
+
+      if (!mustSplit) {
+        //
+        // priority 2: if no stop detection, test for distance
+        // If the distance to the previous location is above distanceThreshold, split
+        //
+        if (GeoTimeSerie.NO_LOCATION != previousLocation && GeoTimeSerie.NO_LOCATION != location && GeoXPLib.orthodromicDistance(location, previousLocation) > distanceThreshold) {
+          splitReason = SPLIT_TYPE_DISTANCE;
+          mustSplit = true;
+        } else if (timestamp - previousTimestamp > timeThreshold) {
+          //
+          // priority 3: if no stop detection and no distance detection, test time split.
+          // If the previous tick was more than 'timeThreshold' ago, split now
+          //
+          splitReason = SPLIT_TYPE_TIME;
+          mustSplit = true;
         }
       }
-      
+
+      //
+      // mustSplit is also true on the first iteration
+      //
       if (mustSplit) {
+        //
+        // End the previous split (could be null on the first iteration), add optional labels.
+        //
+        if (null != splitTypeLabel && null != split) {
+          split.getMetadata().putToLabels(splitTypeLabel, splitReason);
+        }
+        //
+        // Split again the current split if proximityZoneSplit is set
+        //
+        if (proximityZoneSplit && null != split && splitReason == SPLIT_TYPE_STOPPED) {
```

Comment:
Use SPLIT_TYP_STOPPED.equals(splitReason), in your case this is equivalent, but comparing strings using == is errorprone most of the time